### PR TITLE
Support unnamed requirements in `uv add`

### PR DESCRIPTION
**Pull request by ibraheemdev**
## Summary

Support unnamed URL requirements in `uv add`. For example, `uv add git+https://github.com/pallets/flask`.

Part of https://github.com/astral-sh/uv/issues/3959.

### CODE REVIEW

File: crates/uv/src/commands/project/add.rs [hunk 1]
```diff
@@ -33,10 +40,93 @@ pub(crate) async fn add(
     // Find the project requirements.
     let project = ProjectWorkspace::discover(&std::env::current_dir()?, None).await?;
 
+    // Discover or create the virtual environment.
+    let venv = project::init_environment(project.workspace(), python.as_deref(), cache, printer)?;
```
Comment (zanieb) [inline crates/uv/src/commands/project/add.rs:44]: Note we should only discover an interpreter not create an environment if syncing is disabled (looks like there's no toggle for that yet)

File: crates/uv/src/commands/project/add.rs [hunk 2]
```diff
@@ -33,10 +40,93 @@ pub(crate) async fn add(
...
+
+    // Use the default settings.
+    let settings = ResolverSettings::default();
+
+    // Determine the environment for the resolution.
+    let (tags, markers) =
+        resolution_environment(python_version, python_platform, venv.interpreter())?;
+
+    // Initialize the registry client.
+    let client = RegistryClientBuilder::new(cache.clone())
```
Comment (zanieb) [inline crates/uv/src/commands/project/add.rs:73]: Created #4330 to track the duplication of this after declaring a `BaseClientBuilder` above.

File: crates/uv/src/settings.rs [hunk 3]
```diff
@@ -347,20 +347,27 @@ impl LockSettings {
...
+            build,
+            refresh,
             python,
         } = args;
 
         Self {
             requirements,
             python,
+            refresh: Refresh::from(refresh),
+            settings: ResolverSettings::combine(resolver_options(resolver, build), filesystem),
```
Comment (zanieb) [inline crates/uv/src/settings.rs:370]: Since `add` performs a sync shouldn't it receive `ResolverInstallerSettings`?

Comment (zanieb) [reply crates/uv/src/settings.rs:370]: Partially unrelated to your work here. cc @charliermarsh

**Review by zanieb — APPROVED** [review]

File: crates/uv/src/settings.rs (hunk 3 above)
Comment (ibraheemdev) [reply crates/uv/src/settings.rs:370]: Oh yes, I think it should.

**Pull request merged**